### PR TITLE
Feature/subscriptions

### PR DESCRIPTION
**Pull request by partydragen**
Add subscription system

### CODE REVIEW

File: upload/custom/panel_templates/Default/store/product.tpl [hunk 1]
```diff
@@ -118,6 +118,14 @@
                             </div>
                             <div class="form-group">
                                 <div class="row">
+                                    <div class="col-md-6">
+                                        <label for="inputPaymentType">{$RECURRING_PAYMENT}</label>
+                                        <select name="payment_type" class="form-control" id="inputPaymentType">
+                                            <option value="1" {if $RECURRING_PAYMENT_VALUE == 1} selected{/if}>Only charge customer once</option>
+                                            <option value="2" {if $RECURRING_PAYMENT_VALUE == 2} selected{/if}>Charge a recurring subscription</option>
+                                            <option value="3" {if $RECURRING_PAYMENT_VALUE == 3} selected{/if}>Allow one off charge & recurring subscription</option>
```
Comment (supercrafter100) [inline upload/custom/panel_templates/Default/store/product.tpl:126]: Untranslated strings

File: upload/custom/panel_templates/Default/store/product_new.tpl [hunk 2]
```diff
@@ -104,6 +104,14 @@
                             </div>
                             <div class="form-group">
                                 <div class="row">
+                                    <div class="col-md-6">
+                                        <label for="inputPaymentType">{$RECURRING_PAYMENT}</label>
+                                        <select name="payment_type" class="form-control" id="inputPaymentType">
+                                            <option value="1" {if $RECURRING_PAYMENT_VALUE == 1} selected{/if}>Only charge customer once</option>
+                                            <option value="2" {if $RECURRING_PAYMENT_VALUE == 2} selected{/if}>Charge a recurring subscription</option>
+                                            <option value="3" {if $RECURRING_PAYMENT_VALUE == 3} selected{/if}>Allow one off charge & recurring subscription</option>
```
Comment (supercrafter100) [inline upload/custom/panel_templates/Default/store/product_new.tpl:112]: Same as above

File: upload/custom/panel_templates/Default/store/product_new.tpl [hunk 3]
```diff
@@ -104,6 +104,14 @@
...
+                                        </select>
+                                    </div>
                                     <div class="col-md-6">
                                         <label for="inputDurability">Remove from customer after (Expire)</label>
```
Comment (supercrafter100) [inline upload/custom/panel_templates/Default/store/product_new.tpl:116]: Not part of the PR but as it shows up, untranslated

File: upload/modules/Store/classes/StoreConfig.php [hunk 4]
```diff
@@ -1,89 +1,190 @@
...
+/**
+ * Provides static methods to get and set configuration values from the `modules/Store/config.php` file.
  *
- *  License: MIT
- *
- *  Store module
+ * @package NamelessMC\Core
+ * @author Samerton
+ * @version 2.0.0
+ * @license MIT
```
Comment (supercrafter100) [inline upload/modules/Store/classes/StoreConfig.php:8]: This requires updating I think?

File: upload/modules/Store/classes/Subscription.php [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    /**
+     * Update a subscription data in the database.
+     *
+     * @param array $fields Column names and values to update.
+     */
+    public function update(array $fields = []): void {
+        if (!$this->_db->update('store_subscriptions', $this->data()->id, $fields)) {
+            throw new Exception('There was a problem updating product');
```
Comment (supercrafter100) [inline upload/modules/Store/classes/Subscription.php:40]: You're updating a subscription here, not a product?

File: upload/modules/Store/gateways/PayPalBusiness/gateway.php [hunk 6]
```diff
@@ -29,34 +29,71 @@ public function processOrder(Order $order): void {
...
+            } catch (\PayPal\Exception\PayPalConnectionException $ex) {
+                ErrorHandler::logCustomError($ex->getData());
+            }
+        } else {
+            // Payment subscription
+            $product = null;
+            foreach ($order->getProducts() as $item) {
+                $product = $item;
+                break;
+            }
```
Comment (supercrafter100) [inline upload/modules/Store/gateways/PayPalBusiness/gateway.php:71]: What's the idea behind this? Isn't the products list meant to be a single value in this case? Just use $order->getProducts()[0] then?

File: upload/modules/Store/gateways/PayPalBusiness/gateway.php [hunk 7]
```diff
@@ -29,34 +29,71 @@ public function processOrder(Order $order): void {
...
+                $agreement = $agreement->create($apiContext);
 
-            Redirect::to($payment->getApprovalLink());
-        } catch (\PayPal\Exception\PayPalConnectionException $ex) {
-            ErrorHandler::logCustomError($ex->getData());
+                $approvalUrl = $agreement->getApprovalLink();
+
+            } catch (Exception $ex) {
+                echo "Failed to get activate";
+                exit();
```
Comment (supercrafter100) [inline upload/modules/Store/gateways/PayPalBusiness/gateway.php:92]: Just use `return`;

File: upload/modules/Store/gateways/PayPalBusiness/gateway.php [hunk 8]
```diff
@@ -29,34 +29,71 @@ public function processOrder(Order $order): void {
...
+            }
+
+            header("Location:" . $approvalUrl);
```
Comment (supercrafter100) [inline upload/modules/Store/gateways/PayPalBusiness/gateway.php:95]: Use `Redirect::to()` instead

File: upload/modules/Store/gateways/PayPalBusiness/gateway.php [hunk 9]
```diff
@@ -29,34 +29,71 @@ public function processOrder(Order $order): void {
...
+            exit();
```
Comment (supercrafter100) [inline upload/modules/Store/gateways/PayPalBusiness/gateway.php:96]: Use `return;` or just nothing at all, seeing as it's at the end anyways this seems redundant?

File: upload/modules/Store/gateways/PayPalBusiness/gateway.php [hunk 10]
```diff
@@ -137,14 +174,29 @@ public function handleReturn(): bool {
                     $next_billing_date = date("U", strtotime($agreement->getAgreementDetails()->getNextBillingDate()));
                 }
 
+                $payment_definitions = $agreement->getPlan()->getPaymentDefinitions();
+                $payment_definitions = $payment_definitions[0];
+
+                $order_id = $_SESSION['shopping_cart']['order_id'];
+                if ($order_id == null || !is_numeric($order_id)) {
+                    die('INVALID ORDER ID');
```
Comment (supercrafter100) [inline upload/modules/Store/gateways/PayPalBusiness/gateway.php:182]: Would it make more sense to throw an exception instead?

File: upload/modules/Store/gateways/PayPalBusiness/gateway.php [hunk 11]
```diff
@@ -253,17 +324,23 @@ public function handleListener(): void {
 
                             break;
                         case 'BILLING.SUBSCRIPTION.CREATED':
-                            $id = $response->resource->id;
+                            // Subscription created
+                            $subscription = new Subscription($response->resource->id, 'agreement_id');
+                            if ($subscription->exists()) {
+                                $subscription->update([
+                                    'status_id' => 1
```
Comment (supercrafter100) [inline upload/modules/Store/gateways/PayPalBusiness/gateway.php:331]: I feel like using an enum would be beneficial here to make it clearer what exactly the status is. Using a number is very vague and hard to remember

File: upload/modules/Store/gateways/PayPalBusiness/gateway.php [hunk 12]
```diff
@@ -332,11 +409,11 @@ private function getApiContext() {
                         'log.LogEnabled' => true,
                         'log.FileName' => ROOT_PATH . '/cache/logs/PayPal.log',
                         'log.LogLevel' => 'FINE',
-                        'mode' => 'live',
+                        'mode' => 'sandbox',
```
Comment (supercrafter100) [inline upload/modules/Store/gateways/PayPalBusiness/gateway.php:412]: Make sure to revert this! You are only in a sandbox during development!

File: upload/modules/Store/pages/panel/subscriptions.php [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+                'next_billing_date' => date(DATE_FORMAT, $subscription->data()->next_billing_date),
+                'amount_format' => Output::getPurified(
+                    Store::formatPrice(
+                        $subscription->data()->amount_cents,
+                        $subscription->data()->currency,
+                        Store::getCurrencySymbol(),
+                        STORE_CURRENCY_FORMAT,
+                    )
+                ),
+                'link' => URL::build('/panel/store/subscriptions/', 'subscription=' . urlencode($subscription->data()->id))
```
Comment (supercrafter100) [inline upload/modules/Store/pages/panel/subscriptions.php:69]: Surely the id doesn't need to be encoded? It's a number anyways?

File: upload/modules/Store/pages/panel/subscriptions.php [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    // Payments
+    $payments_list = [];
+    $payments_query = DB::getInstance()->query('SELECT * FROM nl2_store_payments WHERE subscription_id = ? ORDER BY created DESC', [$subscription->data()->id]);
+    if ($payments_query->count()) {
+        foreach ($payments_query->results() as $payment) {
+            $payments_list[] = [
+                'date' => date(DATE_FORMAT, $payment->created),
+                'link' => URL::build('/panel/store/payments', 'payment=' . urlencode($payment->id))
```
Comment (supercrafter100) [inline upload/modules/Store/pages/panel/subscriptions.php:146]: Again, redundant urlencode?

**Review by supercrafter100 — CHANGES_REQUESTED** [review]

File: upload/modules/Store/gateways/PayPalBusiness/gateway.php (hunk 6 above)
Comment (partydragen) [reply upload/modules/Store/gateways/PayPalBusiness/gateway.php:71]: Can't do 0 as the key is the product id

Comment (supercrafter100) [reply upload/modules/Store/gateways/PayPalBusiness/gateway.php:71]: Then use something like:
```php
$product = array_values($order-getProducts())[0]
```
or 
```php
$product = reset($order-getProducts())
```
